### PR TITLE
Add support for optional parameters

### PR DESCRIPTION
**Pull request by alirezanet**
This PR adds `IsRequired` to `CommandParameter`, but only can be false if the parameter is the last one.
Analyzer will check this behavior

Related to #99 #76

### CODE REVIEW

File: CliFx.Analyzers/ParameterMustBeRequiredExceptLastOneAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        SyntaxNodeAnalysisContext context,
+        PropertyDeclarationSyntax propertyDeclaration,
+        IPropertySymbol property)
+    {
+        if (property.ContainingType is null)
+            return;
+
+        var parameter = CommandParameterSymbol.TryResolve(property);
+        
+        if (parameter is null || parameter.IsRequired is null or true )
```
Comment (Tyrrrz) [inline CliFx.Analyzers/ParameterMustBeRequiredExceptLastOneAnalyzer.cs:31]:
```suggestion
        if (parameter is null || parameter.IsRequired != false)
```

File: CliFx.Analyzers/ObjectModel/CommandParameterSymbol.cs [hunk 2]
```diff
@@ -10,19 +10,22 @@ internal partial class CommandParameterSymbol
...
+    
+    public bool? IsRequired { get; }
+    
     public ITypeSymbol? ConverterType { get; }
 
     public IReadOnlyList<ITypeSymbol> ValidatorTypes { get; }
 
-    public CommandParameterSymbol(
-        int order,
+    public CommandParameterSymbol(int order,
```
Comment (Tyrrrz) [inline CliFx.Analyzers/ObjectModel/CommandParameterSymbol.cs:20]:
```suggestion
    public CommandParameterSymbol(
        int order,
```

File: CliFx.Analyzers/ObjectModel/CommandParameterSymbol.cs [hunk 3]
```diff
@@ -63,17 +72,14 @@ private static CommandParameterSymbol FromAttribute(AttributeData attribute)
...
 
     public static CommandParameterSymbol? TryResolve(IPropertySymbol property)
     {
         var attribute = TryGetParameterAttribute(property);
 
-        if (attribute is null)
-            return null;
-
-        return FromAttribute(attribute);
+        return attribute is null ? null : FromAttribute(attribute);
```
Comment (Tyrrrz) [inline CliFx.Analyzers/ObjectModel/CommandParameterSymbol.cs:82]:
```suggestion
        return attribute is not null
            ? FromAttribute(attribute)
            : null;
```

File: CliFx/Attributes/CommandParameterAttribute.cs [hunk 4]
```diff
@@ -22,6 +22,13 @@ public sealed class CommandParameterAttribute : Attribute
     /// Only one non-scalar parameter is allowed in a command.
     /// </remarks>
     public int Order { get; }
+    
+    /// <summary>
+    /// Whether this parameter is required. (default: true)
+    /// If an parameter is required, the user will get an error if they don't set it.
```
Comment (Tyrrrz) [inline CliFx/Attributes/CommandParameterAttribute.cs:28]:
```suggestion
    /// If a parameter is required, the user will get an error if they don't set it.
```

File: CliFx/Attributes/CommandParameterAttribute.cs [hunk 5]
```diff
@@ -22,6 +22,13 @@ public sealed class CommandParameterAttribute : Attribute
...
+    /// Only the last CommandParameter can be optional.
```
Comment (Tyrrrz) [inline CliFx/Attributes/CommandParameterAttribute.cs:29]:
```suggestion
    /// Only the last parameter in order can be optional.
```

File: CliFx/CommandBinder.cs [hunk 6]
```diff
@@ -272,12 +272,12 @@ private void BindParameters(CommandInput commandInput, CommandSchema commandSche
             );
         }
 
-        if (remainingParameterSchemas.Any())
+        if (remainingRequiredParameterSchemas.Any())
         {
             throw CliFxException.UserError(
                 "Missing parameter(s):" +
```
Comment (Tyrrrz) [inline CliFx/CommandBinder.cs:278]:
```suggestion
                "Missing required parameter(s):" +
```

File: CliFx/Formatting/HelpConsoleFormatter.cs [hunk 7]
```diff
@@ -157,7 +157,8 @@ private void WriteCommandParameters()
 
         foreach (var parameterSchema in _context.CommandSchema.Parameters.OrderBy(p => p.Order))
         {
-            Write(ConsoleColor.Red, "* ");
+            var star = parameterSchema.IsRequired ? "* " : "  ";
+            Write(ConsoleColor.Red, star);
```
Comment (Tyrrrz) [inline CliFx/Formatting/HelpConsoleFormatter.cs:161]:
```suggestion
            if (parameterSchema.IsRequired)
            {
                Write(ConsoleColor.Red, "* ");
            }
            else
            {
                WriteHorizontalMargin();
            }
```

File: CliFx/Schema/ParameterSchema.cs [hunk 8]
```diff
@@ -14,27 +14,32 @@ internal partial class ParameterSchema : IMemberSchema
...
         Name = name;
         Description = description;
+        IsRequired = isRequired;
         ConverterType = converterType;
         ValidatorTypes = validatorTypes;
     }
 
+  
+
     public string GetFormattedIdentifier() => Property.IsScalar()
```
Comment (Tyrrrz) [inline CliFx/Schema/ParameterSchema.cs:43]:
```suggestion
    public string GetFormattedIdentifier() => Property.IsScalar()
```

File: CliFx.Analyzers/ParameterMustBeRequiredExceptLastOneAnalyzer.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace CliFx.Analyzers;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class ParameterMustBeRequiredExceptLastOneAnalyzer : AnalyzerBase
+{
+    public ParameterMustBeRequiredExceptLastOneAnalyzer()
```
Comment (Tyrrrz) [inline CliFx.Analyzers/ParameterMustBeRequiredExceptLastOneAnalyzer.cs:14]: Also rename file and test class
(for consistency with the non-scalar analyzer)

```suggestion
public class ParameterMustBeLastIfNotRequiredAnalyzer : AnalyzerBase
{
    public ParameterMustBeLastIfNotRequiredAnalyzer()
```

**Review by Tyrrrz — CHANGES_REQUESTED** [review]

**Comment by alirezanet** [comment]
Error is related to my other PL. 

I'll send another PL to fix the line break issue on Linux:

image

File: CliFx/Formatting/HelpConsoleFormatter.cs [hunk 10]
```diff
@@ -157,7 +157,14 @@ private void WriteCommandParameters()
...
         {
-            Write(ConsoleColor.Red, "* ");
+            if (parameterSchema.IsRequired)
+            {
+                Write(ConsoleColor.Red, "* ");
+            }
+            else
+            {
+                WriteHorizontalMargin();
+            }
```
Comment (Tyrrrz) [inline CliFx/Formatting/HelpConsoleFormatter.cs:167]:
```suggestion
            }

```

**Review by Tyrrrz — APPROVED** [review]

**Comment by Tyrrrz** [comment]
Thanks!

**Pull request merged**